package.json

```diff
@@ -16,6 +16,7 @@
     "start": "nwb serve-react-app"
   },
   "dependencies": {
+    "@mehra/ts": "^4.2.2",
```

Comment:
i'm hesitant to introduce a dependency used for a single type!

Reply:
This is not for a type; this is for the implementation of the network JSON parser. It also implements the types Async, Assert, etc which were coded by hand into this project in recently merged PRs, and will remove the overhead from those

Reply:
The dependency has grown a bit since the version imported here, but the documentation remains the valid: https://pihart.github.io/ts/classes/network.html#loadjson

Pure typescript code (types and interfaces) has a purple `T` icon (types) or a green wireframe icon (intefaces). The rest, including this JSON parser and the error classes it returns, has some javascript emit

Reply:
loading from network json really shouldnt be this complicated? isnt this just like a two liner in vanilla JS

Reply:
Well you saw the previous implementation in #103. If you use the XHR API, you have to set a bunch of callbacks, and then wrap it in a Promise wrapper ([dependency's implementation](https://github.com/pihart/ts/blob/9e1bb672d6e71e9e95bebbf0f4bc47ca9c97c847/src/lib/network.ts#L41-L92)). Some more (trivial-ish) code is needed to make the thrown errors have a semantic meaning ([dependency's implementation](https://github.com/pihart/ts/blob/9e1bb672d6e71e9e95bebbf0f4bc47ca9c97c847/src/lib/network.ts#L4-L14)).

It is shorter, if you build off of the fetch API, but I don't like that for quite a few reasons:
* XHR has more support
* The Promise from fetch resolves no matter what; even if the response has code in 400-599. You would then have to write a lambda as a `then` which throws its own error based on a given condition (before any functional handling).
* XHR exposes more features

I also agree with what you've said in the past that qboard should not be managing network.

Reply:
i am still leaning on avoiding the dependency. i don't see why there would need to be that much boilerplate above a fetch call?

Reply:
There's not much more I can say beyond the previous comment:
- You get the utility types
  - Can add strictness
  - Can write type tests (these are surprisingly useful when you write your own interfaces)
- You don't have to worry about *any* of the network logic or implementation. For the most part, you can also ignore the API (but not the standard) on which it operates
  - You are not responsible for keeping the client-side network requests up to date with latest technologies
  - You keep the code clean of *any* boilerplate.
    - It becomes easy and safe to mock out the network request in tests
- You get ready-made semantic error classes

Regarding the actual existence of boilerplate, this is what the fetch itself must look like:
* If you inline the code in `qboard.ts`
    ```ts
    if (queryParams.get("json") !== null)
      window
        .fetch(queryParams.get("json"))
        .then((response) => {
          if (response.ok) return response;
          throw new Error(
            "This should actually throw the semantic error class"
          );
          // throw new ResourceNotFoundException(
          //   response.status,
          //   response.statusText
          // );
        })
        .then((response) => response.json())
        .then(JSONReader.readParsed)
        .then(this.pages.overwritePages)
        .catch(console.error);
     ```
* If you extract it ([sourse](https://github.com/pihart/ts/blob/main/src/lib/network.ts#L112-L140))
    ```ts
    // --- https://github.com/pihart/ts/blob/main/src/lib/network.ts#L112-L140

    /**
     * Makes a network request to a resource and returns its data parsed as JSON
     *
     * Use instead of `window.fetch` because it rejects failed transactions.
     *
     * @return The result of the request to the resource, parsed as JSON, if it is `ok`
     * @throws [[`ResourceNotFoundException`]] if not `ok`
     */
    const loadJSON = async (
      input: RequestInfo,
      init?: RequestInit
    ): Promise<any> => {
      const response = await fetchWithError(input, init);
      return response.json();
    }
  
  
    /**
     * Makes a network request
     *
     * Use instead of `window.fetch` because it rejects failed transactions.
     *
     * @return The result of `window.fetch`, if it is `ok`
     * @throws [[`ResourceNotFoundException`]] if not `ok`
     */
   // Note that I rename this from fetch so that I don't declare two variables in window with the same name
    const fetchWithError = async (
      input: RequestInfo,
      init?: RequestInit
    ): Promise<Response> => {
      const response = await window.fetch(input, init);
      if (response.ok) return response;
      throw new ResourceNotFoundException(response.status, response.statusText);
    }
    ```
and this is what you need to add if you want semantic errors (source: [A](https://github.com/pihart/ts/blob/9e1bb672d6e71e9e95bebbf0f4bc47ca9c97c847/src/lib/error.ts#L1-L29), [B](https://github.com/pihart/ts/blob/37a14bce936e201fdc61bf47af97ec05a9ecb323/src/lib/network.ts#L4-L14)):
```ts
// --- https://github.com/pihart/ts/blob/9e1bb672d6e71e9e95bebbf0f4bc47ca9c97c847/src/lib/error.ts#L1-L29

/**
 * Sets a custom name for `Error` via the `constructor` property.
 * Extend this if you want custom names in errors.
 *
 * @warn If you have a minifier that minifies the name of the extending class,
 * this will set the name to the minified name.
 */
export class CustomError extends Error {
  /**
   * @param message The error message
   */
  constructor(message?: string) {
    super(message);
    this.name = this.constructor.name;
  }
}

/**
 * {@linkcode CustomError} which stores `public code: number`
 */
export class ErrorWithCode extends CustomError {
  /**
   * @param code The (public) error code
   * @param message The error message
   */
  constructor(public code: number, message?: string) {
    super(message);
  }
}

// --- https://github.com/pihart/ts/blob/37a14bce936e201fdc61bf47af97ec05a9ecb323/src/lib/network.ts#L4-L14

export class HTTPError extends ErrorWithCode {
  constructor(code = 400, message = "HTTP Error") {
    super(code, message);
  }
}


export class ResourceNotFoundException extends HTTPError {
  constructor(code = 404, message = "Resource not found") {
    super(code, message);
  }
}
```

This is as opposed to
```ts
if (queryParams.get("json") !== null)
  Network.loadJSON(queryParams.get("json"))
    .then(JSONReader.readParsed)
    .then(this.pages.overwritePages)
    .catch(console.error);
```
with the library.

So I agree that—for the fetch itself—there's not *too* much boilerplate if you write it by hand (although only if done directly in `qboard.ts`, which I believe is bad style), but I do think that the semantic errors, simpler usage (truly a one-liner), separations of concerns, and utility types make it overall worth it to include the dependency.